### PR TITLE
Improve viam robot logs cli

### PR DESCRIPTION
**Pull request by zaporter-work**
This PR adds two improvements to the `viam robot logs` command.

1) It reverses the order that the logs are printed so that they are printed in oldest -> newest fashion. This was very confusing when trying to debug and things were happening in an order that seemed impossible. App.viam.com can get away with ordering the logs in reverse, however on a cli, it feels very strange to have reversed log order. (time I wasted before realizing this: 20 minutes)

2) It parses the log Fields entry. Without parsing this, it is nearly impossible to debug certain error messages.

Error prior to this PR:
```
	2024-02-27T02:51:45.626Z	error	robot_server.rdk:service:vision/testv	service config error; starting robot without service
```

I have no idea what the error message is....

Error after this PR:
```
	2024-02-27T02:51:45.626Z	error	robot_server.rdk:service:vision/testv	service config error; starting robot without service	name testv	error mlmodel_name cannot be empty
```

This is rendered this way because this is how we render the fields in app. 

(time I wasted before realizing this: 30 minutes)

Justification for not adding tests: Adding tests won't catch any regressions because any regressions will come from changes in the Fields struct which we don't have a type for. Also, anyone changing this will end up just having to run logs manually anyways. 

If any of you would rather do this yourselves (with a ticket), feel free to say so and we can close this and move to a new PR

-------

If you want to try this yourself, this robot on viam-dev has some logs:
```viam robot logs --robot e4713ae5-013a-43fe-800e-ff7959a8e3a0```

### CODE REVIEW

File: cli/client.go [hunk 1]
```diff
@@ -920,15 +921,24 @@ func (c *viamClient) robotParts(orgStr, locStr, robotStr string) ([]*apppb.Robot
...
-	for _, log := range logs {
+	// Iterate over logs in reverse because they are returned in
+	// order of latest to oldest but we should print from oldest -> newest
+	for i := len(logs) - 1; i >= 0; i-- {
+		log := logs[i]
+		fields, err := parseLogEntryFields(log.Fields)
+		if err != nil {
+			warningf(c.c.App.ErrWriter, "%v", err)
+		}
+		fieldsString := logEntryFieldsToString(fields)
```
Comment (zaporter-work) [inline cli/client.go:932]: This func is fine if fields == nil

File: cli/client.go [hunk 2]
```diff
@@ -920,15 +921,24 @@ func (c *viamClient) robotParts(orgStr, locStr, robotStr string) ([]*apppb.Robot
...
 		printf(
 			c.c.App.Writer,
-			"%s%s\t%s\t%s\t%s",
+			"%s%s\t%s\t%s\t%s%s",
```
Comment (zaporter-work) [inline cli/client.go:935]: no \t here because it is added in the conversion to fields string func

**Comment by zaporter-work** [comment]
Talked with benji offline and added parsing for the Interface field of the zap logs
``` 2024-02-24T00:23:05.877Z	error	robot_server.rdk:component:camera/faketlemilytest	resource build error: dependency "test" is not ready yet; reason=resource "rdk:component:camera/test" not available; reason="resource build error: failed to find camera: found no webcams: failed to find the best driver that fits the constraints"	resource map[API:map[Type:map[namespace:rdk type:component] subtype:camera] Name:faketlemilytest Remote:]	model map[Family:map[model_family:timelapse namespace:zaporter] name:v1-fake]```

File: cli/client.go [hunk 3]
```diff
@@ -1197,3 +1209,45 @@ func (c *viamClient) startRobotPartShell(
...
+		return nil, err
+	}
+	return parsedFields, nil
+}
+
+func logEntryFieldsToString(fields []zap.Field) string {
+	fieldsString := ""
+	for _, field := range fields {
+		fieldValue := ""
+		// TODO use zap encoder to encode
```
Comment (benjirewis) [inline cli/client.go:1229]: Would you mind filing a ticket for this and I can eventually use some yet-to-be-created `EncodeZapFields` utility function in the module logging field parsing as well?

File: cli/client.go [hunk 4]
```diff
@@ -1197,3 +1209,45 @@ func (c *viamClient) startRobotPartShell(
...
+		// it is a bit hard to do that though because sending the data over the wire
+		// loses type info which causes the zap encoders to panic.
+		// This code is modeled after zapcore.Field.AddTo
+		//nolint:exhaustive
```
Comment (benjirewis) [inline cli/client.go:1233]: Needed with the `default` case?

File: cli/client.go [hunk 5]
```diff
@@ -1197,3 +1209,45 @@ func (c *viamClient) startRobotPartShell(
...
+		case zapcore.Int64Type, zapcore.Int32Type, zapcore.Int16Type, zapcore.Int8Type:
+			fieldValue = fmt.Sprint(field.Integer)
+		case zapcore.Float64Type:
+			fieldValue = fmt.Sprint(math.Float64frombits(uint64(field.Integer)))
+		case zapcore.Float32Type:
+			fieldValue = fmt.Sprint(math.Float32frombits(uint32(field.Integer)))
+		default:
+			fieldValue = fmt.Sprint(field.Interface)
+		}
+		fieldsString += fmt.Sprintf("\t%s %s", field.Key, fieldValue)
```
Comment (benjirewis) [inline cli/client.go:1250]: @dgottlieb do we prefer tab-separated key-value pairs like this? Or, something like `{"key": "value", "key2": 1}` instead? I'm fine with either, but want to align how we print fields for the CLI, modules and app.

**Review by benjirewis — COMMENTED** [review]
Nice! I see you have discovered the joy that is `zap.Field`.

File: cli/client.go (hunk 5 above)
Comment (dgottlieb) [reply cli/client.go:1250]: We should go with `{"key": "value", "key2": 1}`.

Comment (cheukt) [reply cli/client.go:1250]: ^ agree with that

File: cli/client.go (hunk 4 above)
Comment (zaporter-work) [reply cli/client.go:1233]: I don't think we have `defaultSignifiesExhaustive` set on our linter (defaults to false) https://golangci-lint.run/usage/linters/#exhaustive

Comment (benjirewis) [reply cli/client.go:1233]: That'd be it! I like leaving it as false; thanks for looking that up.

File: cli/client.go [hunk 6]
```diff
@@ -1197,3 +1209,45 @@ func (c *viamClient) startRobotPartShell(
...
+		case zapcore.StringType, zapcore.ErrorType:
+			fieldValue = field.String
+		case zapcore.BoolType:
+			fieldValue = fmt.Sprint(field.Integer == 1)
+		case zapcore.DurationType:
+			fieldValue = time.Duration(field.Integer).String()
+		case zapcore.Int64Type, zapcore.Int32Type, zapcore.Int16Type, zapcore.Int8Type:
+			fieldValue = fmt.Sprint(field.Integer)
+		case zapcore.Float64Type:
+			fieldValue = fmt.Sprint(math.Float64frombits(uint64(field.Integer)))
```
Comment (mattjperez) [inline cli/client.go:1244]: Does this conversion return the bit-values or the interpreted value?
I'm not sure how the whole `float_to_log` -> `zap.Field.Int64` -> `Uint64` -> `Float64` conversion works in our setup.

Comment (benjirewis) [reply cli/client.go:1244]: I'll talk about this in the NetCode meeting today. The conversion is unfortunately both complicated and lossy from what I understand.

Comment (benjirewis) [reply cli/client.go:1244]: That SGTM; for posterity we're referring to https://github.com/viamrobotics/rdk/pull/3603.

File: cli/client.go (hunk 3 above)
Comment (zaporter-work) [reply cli/client.go:1229]: Deferring till https://github.com/viamrobotics/rdk/pull/3603.

Will swap to that.

**Comment by zaporter-work** [comment]
This robot has some logs with errors:
`viam robot logs --robot caac846f-3a11-4743-834b-69472dca7088`
```
	2024-02-26T17:41:55.551Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = upstream connect error or disconnect/reset before headers. reset reason: connection termination"}
	2024-02-26T22:37:25.738Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = upstream connect error or disconnect/reset before headers. reset reason: connection termination"}
	2024-02-29T00:46:49.358Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = closing transport due to: connection error: desc = \"error reading from server: EOF\", received prior goaway: code: NO_ERROR, debug data: \"server_shutting_down\""}
	2024-02-29T15:45:45.803Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = closing transport due to: connection error: desc = \"error reading from server: EOF\", received prior goaway: code: NO_ERROR, debug data: \"server_shutting_down\""}
	2024-02-29T15:52:58.338Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = closing transport due to: connection error: desc = \"error reading from server: EOF\", received prior goaway: code: NO_ERROR, debug data: \"server_shutting_down\""}
	2024-02-29T16:06:26.367Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = closing transport due to: connection error: desc = \"error reading from server: EOF\", received prior goaway: code: NO_ERROR, debug data: \"server_shutting_down\""}
	2024-02-29T16:11:45.006Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = error reading from server: EOF"}
	2024-02-29T16:20:15.801Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = error reading from server: EOF"}
	2024-02-29T16:29:13.719Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = closing transport due to: connection error: desc = \"error reading from server: EOF\", received prior goaway: code: NO_ERROR, debug data: \"server_shutting_down\""}
	2024-02-29T16:40:21.233Z	error	robot_server.external_signaler	error answering	{"error": "rpc error: code = Unavailable desc = closing transport due to: connection error: desc = \"error reading from server: EOF\", received prior goaway: code: NO_ERROR, debug data: \"server_shutting_down\""}
```

File: cli/client.go (hunk 5 above)
Comment (zaporter-work) [reply cli/client.go]: Fixed in [77a0c1f](https://github.com/viamrobotics/rdk/pull/3613/commits/77a0c1f4daf40e9f7c4300e0799a21e4d4e01909)

File: cli/client.go [hunk 7]
```diff
@@ -939,15 +940,24 @@ func (c *viamClient) robotParts(orgStr, locStr, robotStr string) ([]*apppb.Robot
...
+		if err != nil {
+			warningf(c.c.App.ErrWriter, "%v", err)
+			fieldsString = ""
+		}
 		printf(
 			c.c.App.Writer,
-			"%s%s\t%s\t%s\t%s",
+			"%s%s\t%s\t%s\t%s\t%s",
 			indent,
 			log.Time.AsTime().Format("2006-01-02T15:04:05.000Z0700"),
```
Comment (benjirewis) [inline cli/client.go:956]: Mind changing this to `logging.DefaultTimeFormatStr` (it's the same but a `const`)?

File: cli/client_test.go [hunk 8]
```diff
@@ -218,3 +221,40 @@ func TestBaseURLParsing(t *testing.T) {
...
+		f2, err := logging.FieldToProto(zapcore.Field{
+			Key:     "key2",
+			Type:    zapcore.Int32Type,
+			Integer: 123,
+		})
+		test.That(t, err, test.ShouldBeNil)
+		f3, err := logging.FieldToProto(zapcore.Field{
+			Key:       "facts",
+			Type:      zapcore.ReflectType,
+			Interface: map[string]string{"app.viam": "cool", "cli": "cooler"},
```
Comment (benjirewis) [inline cli/client_test.go:242]: ❤️

**Review by cheukt — APPROVED** [review]

**Pull request merged**